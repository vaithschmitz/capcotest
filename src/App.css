body {
  margin: 0;
  padding: 0;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.App {
  text-align: center;
}

.leftSide, .rightSide{
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 50%;
  height: 100vh;
  color: blue;
}

.leftSide{
  float: left;
  background-color: rgba(120, 120, 223, 0.644);
  box-shadow: 100px 0px 159px 53px rgba(120, 120, 223, 0.644);
}

.rightSide{
  float: right;
  background-color: rgba(145, 145, 219, 0.644);
  box-shadow: -100px 0px 159px 53px rgba(145, 145, 219, 0.644);
}