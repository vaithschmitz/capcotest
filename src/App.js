import React, { Component } from 'react';
import './App.css'

export class App extends Component {
  constructor(props){
    super(props)
    this.state = {
      GBP: '',
      USD: '',
      EUR: '',
      hasConverted: false,
      hasError: false
    }
    this.handleSubmit = this.handleSubmit.bind(this)
    this.handleChange = this.handleChange.bind(this)
    this.fetchRates = this.fetchRates.bind(this)
    this.handleDisplay = this.handleDisplay.bind(this)
  }

  handleSubmit(e){
    this.fetchRates();
    e.preventDefault()
  }

  handleChange(e){
    this.setState({GBP: e.target.value})
  }

  async fetchRates(){
    let obj = {amount: this.state.GBP}
    const proxyurl = "https://cors-anywhere.herokuapp.com/";

    try {
      let res = await fetch(proxyurl + 'https://us-central1-ornate-apricot-220209.cloudfunctions.net/CodingTestFrontEnd-API?',
        {
          method: 'POST', 
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(obj)
        })

      let data = await res.json()

      this.setState({
        USD: data.USD.toFixed(2),
        EUR: data.EUR.toFixed(2),
        hasConverted: true
      })

    } catch(err) {
      console.log(err)
      this.setState({
        hasConverted: false,
        hasError: true
      })
    }
  }

  handleDisplay(){
    return(
      <div>
        Your Money Will Buy You: 
      <br/>
        {
          this.state.USD > this.state.EUR ? 
            <div> USD {this.state.USD} <br/> EUR {this.state.EUR} <br/>That's {(this.state.USD - this.state.EUR).toFixed(2)} more USD than EUR!</div>  : 
            <div> EUR {this.state.EUR} <br/> USD {this.state.USD} <br/>That's {(this.state.EUR - this.state.USD).toFixed(2)} more EUR than USD!</div>
        } 
      </div>
    )    
  }

  render() {
    console.log(this.state.USD)
    return (
      <div className='App'>

        <div className='leftSide'>
          <form id='GBPform' onSubmit={this.handleSubmit}>
            Enter GBP Amount: 
            <br/>
            <input id='GBPinput' type='number' min='0' step='0.01' value={this.state.GBP} onChange={this.handleChange}></input>
            <br/>
            <input id='GBPsubmit' type='submit' value='Convert'></input>
          </form>
        </div>

        <div className='rightSide'>
          {this.state.hasConverted ? this.handleDisplay() : 
            (this.state.hasError ? "There's Been A Problem. Refresh And Try Again In A Few Minutes" : '')}
        </div>

      </div>
    )
  }
}

export default App
