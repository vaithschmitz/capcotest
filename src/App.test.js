import React from 'react';
import {shallow} from 'enzyme';
import App from './App';


describe('render tests', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<App/>)
    expect(wrapper).toMatchSnapshot()
   });
  
  it('renders input form', () => {
    const wrapper = shallow(<App/>)
    expect(wrapper.exists('#GBPform')).toEqual(true)
  });
});


describe('input state changes', () => {

  it('currency states are empty on launch', () => {
    const wrapper = shallow(<App/>)
    expect(wrapper.state('GBP')).toEqual('')
    expect(wrapper.state('EUR')).toEqual('')
    expect(wrapper.state('USD')).toEqual('')
  });

  it('user input changes GBP state', () => {
    const wrapper = shallow(<App/>)
    const GBPinput = wrapper.find('#GBPinput')
    const form = wrapper.find('#GBPform')

    GBPinput.simulate('change', { target: { value: 14}})
    form.simulate('submit', { preventDefault () {} })
    expect(wrapper.state('GBP')).toBe(14)
  });
});

describe('fetchRates', () => {
  let mockEvent
  let mockResponse
  let mockfetchRates
  let wrapper
  
  beforeEach( () =>{
    mockEvent = { preventDefault: jest.fn() }
    mockResponse = {USD: 18.00, EUR: 17.00};
    mockfetchRates = jest.fn();
    wrapper = shallow(<App fetchRates={mockfetchRates} />)
  })

  it('successful fetch changes state accordingly', done => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve(mockResponse)
    }))  
    wrapper.instance().fetchRates(mockEvent)
    .then(() => {
      expect(wrapper.state('hasConverted')).toBeTruthy()
      expect(wrapper.state('hasError')).toBeFalsy()
      done()
    });
  });

  it('unsuccessful fetch changes state accordingly', done => {
    window.fetch = jest.fn().mockImplementationOnce(() => Promise.reject(
      new Error('Fetch failed')
    ))
  
    wrapper.instance().fetchRates(mockEvent)
    .then(() => {
      expect(wrapper.state('hasConverted')).toBeFalsy()
      expect(wrapper.state('hasError')).toBeTruthy()
      done()
    });
  })

  it('fetch changes USD/EUR state', done => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve(mockResponse)
    })) 

    wrapper.instance().fetchRates(mockEvent)
    .then(() => {
      expect(wrapper.state('USD')).toBe('18.00')
      expect(wrapper.state('EUR')).toBe('17.00')
      done()
    })
  });
});
    


