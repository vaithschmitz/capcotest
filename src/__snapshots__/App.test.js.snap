// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render tests renders correctly 1`] = `ShallowWrapper {}`;
